hw2/hw2.py

```diff
@@ -0,0 +1,420 @@
+        weighted_impurities_children = np.sum(
+            [
+                (child.data.shape[0] / n_total_sample) * child.impurity_func(child.data)
+                for child in self.children
+            ]
+        )
+
+        # Feature importance calculation
+        self.feature_importance = (
+            weighted_impurity_current - weighted_impurities_children
+        )
+
+    def goodness_of_split(self, feature, gain_ratio=False):
+        """
+        Calculate the goodness of split of a dataset given a feature and impurity function.
+
+        Input:
+        - feature: the feature index the split is being evaluated according to.
+
+        Returns:
+        - goodness: the goodness of split
+        - groups: a dictionary holding the data after splitting
+                  according to the feature values.
+        """
+        # Splitting the data by feature values
+        unique_values, indices = np.unique(self.data[:, feature], return_inverse=True)
+        groups = {
+            val: self.data[indices == idx] for idx, val in enumerate(unique_values)
+        }
+
+        # Calculate impurity before the split
+        impurity_before = self.impurity_func(self.data)
+
+        # Calculate weighted impurity after the split
+        n_samples = len(self.data)
+        weighted_impurity_after = sum(
+            (len(group) / n_samples) * self.impurity_func(group)
+            for group in groups.values()
+        )
+
+        # Goodness of Split
+        goodness = impurity_before - weighted_impurity_after
+
+        if gain_ratio:
+            # Calculate split information
+            split_info = -sum(
+                (len(group) / n_samples) * np.log2(len(group) / n_samples)
+                for group in groups.values()
+                if len(group) > 0  # Avoid log2(0) which is undefined
+            )
+            # Gain Ratio
+            if split_info > 0:
+                goodness = goodness / split_info
+            else:
+                goodness = 0  # To handle cases where split_info is zero, avoiding division by zero
+
+        return goodness, groups
+
+    def split(self):
+        """
+        Splits the current node according to the self.impurity_func. This function finds
+        the best feature to split according to and create the corresponding children.
+        This function should support pruning according to self.chi and self.max_depth.
+
+        This function has no return value
+        """
+        if self.depth >= self.max_depth or len(self.data) == 0:
+            self.terminal = True
+            return  # Stop splitting if max depth is reached or no data is present
+
+        # Select the best feature based on the impurity reduction
+        best_goodness = -np.inf
+        best_feature = None
+        best_splits = {}
+
+        for feature in range(self.data.shape[1] - 1):  # Assume last column is the label
+            goodness, splits = self.goodness_of_split(feature, self.gain_ratio)
```

Comment:
After reverting signature of - `goodness_of_split`, no need to pass `self.gain_ratio`.